pkg/flow/internal/controller/node_component.go

```diff
@@ -394,16 +388,6 @@ func (cn *ComponentNode) setExports(e component.Exports) {
 	}
 	cn.exportsMut.Unlock()
 
-	if cn.doingEval.Load() {
-		// Optimization edge case: some components supply exports when they're
-		// being evaluated.
-		//
-		// Since components that are being evaluated will always cause their
-		// dependencies to also be evaluated, there's no reason to call
-		// onExportsChange here.
-		return
-	}
```

Comment:
This is an important change, which allows us to only evaluate direct incoming nodes in the graph.